### PR TITLE
switch from blake2-rfc to blake2b_simd

### PR DESCRIPTION
**Pull request by oconnor663**
This is a faster implementation of BLAKE2b, which takes advantage of AVX2 (and other) SIMD optimizations. Here are the results of hashing a 1 GB file under `hyperfine`. `bin_orig` is a release build of the current master branch, and `bin_simd` is a release build of this PR. `blake2` is the binary provided by the [`blake2_bin`](https://crates.io/crates/blake2_bin) crate, which uses the same `blake2b_simd` implementation as in this PR.

```none
❯ hyperfine "/tmp/bin_orig hashsum --b2sum f" "/tmp/bin_simd hashsum --b2sum f" "blake2 f"
Benchmark #1: /tmp/bin_orig hashsum --b2sum f
  Time (mean ± σ):      1.969 s ±  0.015 s    [User: 1.436 s, System: 0.530 s]
  Range (min … max):    1.943 s …  1.998 s    10 runs
 
Benchmark #2: /tmp/bin_simd hashsum --b2sum f
  Time (mean ± σ):      1.505 s ±  0.002 s    [User: 967.0 ms, System: 535.5 ms]
  Range (min … max):    1.503 s …  1.511 s    10 runs
 
Benchmark #3: blake2 f
  Time (mean ± σ):      1.097 s ±  0.004 s    [User: 983.2 ms, System: 111.9 ms]
  Range (min … max):    1.088 s …  1.101 s    10 runs
 
Summary
  'blake2 f' ran
    1.37 ± 0.00 times faster than '/tmp/bin_simd hashsum --b2sum f'
    1.80 ± 0.01 times faster than '/tmp/bin_orig hashsum --b2sum f'
```

In this test, this change speeds up `hashsum` by ~30%. We can see that `blake2` is faster still, even though its implementation of BLAKE2b is the same as in this PR. I believe this is because of the inefficient IO loop described in https://github.com/uutils/coreutils/issues/2295, but I haven't confirmed this.

### CODE REVIEW

File: src/uu/hashsum/src/digest.rs [hunk 1]
```diff
@@ -49,22 +48,22 @@ impl Digest for md5::Context {
...
+        Self::new()
     }
 
     fn input(&mut self, input: &[u8]) {
         self.update(input);
     }
 
     fn result(&mut self, out: &mut [u8]) {
-        let hash_result = &self.clone().finalize();
+        let hash_result = &self.finalize();
```
Comment (oconnor663) [inline src/uu/hashsum/src/digest.rs:61]: `blake2b_simd::State::finalize` takes `&self` and is idempotent.

**Comment by sylvestre** [comment]
wahou, excellent. Could you please document the benchmark in https://github.com/uutils/coreutils/blob/master/src/uu/cut/BENCHMARKING.md ?
thanks

**Comment by sylvestre** [comment]
@oconnor663   ping ?

**Comment by sylvestre** [comment]
Documented in
https://github.com/uutils/coreutils/pull/2363

**Comment by oconnor663** [comment]
Thank you! Apologies for not getting back to this sooner.

By the way, would you be interested in BLAKE3 support in hashsum? I'd be happy to put together another PR.

**Comment by sylvestre** [comment]
Oh, you are upstream too ;)
Sure. We have 1.43.1 as minimal version for rust btw

**Comment by oconnor663** [comment]
Ok, I need to check the msrv for the blake3 crate. The imminently arriving 1.0 version will depend on const generics, but the current 0.8.3 version will be functionally identical and has a lower msrv. So we could either use that, or just wait until uutils is ready to use const generics.

**Comment by oconnor663** [comment]
(confirmed that blake3 0.3.8 builds under rustc 1.43.1)

**Comment by DestroyerXyz** [comment]
@oconnor663 Are you still working on BLAKE3? If not, I can take over if you want.

**Comment by oconnor663** [comment]
One of them yes :)

I'm happy to let you take charge of BLAKE3 feature for uutils. Hopefully the `blake3` crate API is simple and everything Just Works™, but please let me know if anything gives you trouble.

**Comment by DestroyerXyz** [comment]
Ok thanks

**Pull request merged**